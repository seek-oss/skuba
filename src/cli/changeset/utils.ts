// Adapted from https://github.com/changesets/action/blob/21240c3cd1d2efa2672d64e0235a03cf139b83e6/src/gitUtils.ts

import type { Package } from '@manypkg/get-packages';
import { getPackages } from '@manypkg/get-packages';
import execa from 'execa';
import mdastToString from 'mdast-util-to-string';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import { unified } from 'unified';

export const BumpLevels = {
  dep: 0,
  patch: 1,
  minor: 2,
  major: 3,
} as const;

export async function getVersionsByDirectory(cwd: string) {
  const { packages } = await getPackages(cwd);
  return new Map(packages.map((x) => [x.dir, x.packageJson.version]));
}

export async function getChangedPackages(
  cwd: string,
  previousVersions: Map<string, string>,
) {
  const { packages } = await getPackages(cwd);
  const changedPackages = new Set<Package>();

  for (const pkg of packages) {
    const previousVersion = previousVersions.get(pkg.dir);
    if (previousVersion !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}

export function getChangelogEntry(changelog: string, version: string) {
  const ast = unified().use(remarkParse).parse(changelog);

  let highestLevel: number = BumpLevels.dep;

  const nodes = ast.children;
  let headingStartInfo:
    | {
        index: number;
        depth: number;
      }
    | undefined;
  let endIndex: number | undefined;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (node.type === 'heading') {
      const stringified: string = mdastToString.toString(node);
      const match = /(major|minor|patch)/.exec(stringified.toLowerCase());
      if (match !== null) {
        const level = BumpLevels[match[0] as 'major' | 'minor' | 'patch'];
        highestLevel = Math.max(level, highestLevel);
      }
      if (headingStartInfo === undefined && stringified === version) {
        headingStartInfo = {
          index: i,
          depth: node.depth,
        };
        continue;
      }
      if (
        endIndex === undefined &&
        headingStartInfo !== undefined &&
        headingStartInfo.depth === node.depth
      ) {
        endIndex = i;
        break;
      }
    }
  }
  if (headingStartInfo) {
    ast.children = ast.children.slice(headingStartInfo.index + 1, endIndex);
  }
  return {
    content: unified().use(remarkStringify).stringify(ast),
    highestLevel,
  };
}

export const execWithOutput = async (
  command: string,
  args?: string[],
  options: { ignoreReturnCode?: boolean; cwd?: string } = {},
) => {
  try {
    const { exitCode, stdout, stderr } = await execa(
      command,
      args,
      ...(options.cwd ? [{ cwd: options.cwd }] : []),
    );

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  } catch ({ exitCode, stdout, stderr }) {
    if (!options.ignoreReturnCode) {
      throw new Error(stderr as string);
    }

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  }
};

export function sortTheThings(
  a: { private: boolean; highestLevel: number },
  b: { private: boolean; highestLevel: number },
) {
  if (a.private === b.private) {
    return b.highestLevel - a.highestLevel;
  }
  if (a.private) {
    return 1;
  }
  return -1;
}
