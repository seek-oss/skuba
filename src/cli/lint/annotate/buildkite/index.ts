import * as Buildkite from '../../../../api/buildkite';
import type { ESLintOutput } from '../../../adapter/eslint';
import type { PrettierOutput } from '../../../adapter/prettier';
import type { StreamInterceptor } from '../../../lint/external';
import type { InternalLintResult } from '../../internal';

import { createEslintAnnotations } from './eslint';
import { createInternalAnnotations } from './internal';
import { createPrettierAnnotations } from './prettier';
import { createTscAnnotations } from './tsc';

export const createBuildkiteAnnotations = async (
  internal: InternalLintResult,
  eslint: ESLintOutput,
  prettier: PrettierOutput,
  tscOk: boolean,
  tscOutputStream: StreamInterceptor,
): Promise<void> => {
  if (internal.ok && eslint.ok && prettier.ok && tscOk) {
    return;
  }

  const buildkiteOutput = [
    '`skuba lint` found issues that require triage:',
    ...createInternalAnnotations(internal),
    ...createEslintAnnotations(eslint),
    ...createPrettierAnnotations(prettier),
    ...createTscAnnotations(tscOk, tscOutputStream),
  ].join('\n\n');

  await Buildkite.annotate(buildkiteOutput, {
    context: 'skuba-lint',
    scopeContextToStep: true,
    style: 'error',
  });
};
