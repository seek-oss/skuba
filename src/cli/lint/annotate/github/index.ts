import * as GitHub from '../../../../api/github';
import {
  buildNameFromEnvironment,
  enabledFromEnvironment,
} from '../../../../api/github/environment';
import type { ESLintOutput } from '../../../adapter/eslint';
import type { PrettierOutput } from '../../../adapter/prettier';
import type { StreamInterceptor } from '../../../lint/external';
import type { InternalLintResult } from '../../internal';

import { createEslintAnnotations } from './eslint';
import { createInternalAnnotations } from './internal';
import { createPrettierAnnotations } from './prettier';
import { createTscAnnotations } from './tsc';

export const createGitHubAnnotations = async (
  internal: InternalLintResult,
  eslint: ESLintOutput,
  prettier: PrettierOutput,
  tscOk: boolean,
  tscOutputStream: StreamInterceptor,
) => {
  if (!enabledFromEnvironment()) {
    return;
  }

  const annotations: GitHub.Annotation[] = [
    ...createInternalAnnotations(internal),
    ...createEslintAnnotations(eslint),
    ...createPrettierAnnotations(prettier),
    ...createTscAnnotations(tscOk, tscOutputStream),
  ];

  const isOk = eslint.ok && prettier.ok && internal.ok && tscOk;

  const summary = isOk
    ? '`skuba lint` passed.'
    : '`skuba lint` found issues that require triage.';

  const build = buildNameFromEnvironment();

  await GitHub.createCheckRun({
    name: 'skuba/lint',
    summary,
    annotations,
    conclusion: isOk ? 'success' : 'failure',
    title: `${build} ${isOk ? 'passed' : 'failed'}`,
  });
};
