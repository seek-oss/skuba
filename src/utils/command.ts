import path from 'path';

export type Command = (typeof COMMAND_LIST)[number];

export const COMMAND_ALIASES: Record<string, Command> = {
  '-h': 'help',
  '--help': 'help',
  '-v': 'version',
  '--version': 'version',
};

export const COMMAND_DIR = path.join(__dirname, '..', 'cli');

export const COMMAND_LIST = [
  'build',
  'build-package',
  'configure',
  'format',
  'help',
  'init',
  'lint',
  'node',
  'release',
  'start',
  'test',
  'version',
] as const;

export const COMMAND_SET = new Set<string>(COMMAND_LIST);

export const commandToModule = (command: Command): string =>
  command
    .split('-')
    .map((segment, index) =>
      index === 0
        ? segment
        : `${segment[0].toLocaleUpperCase()}${segment.slice(1)}`,
    )
    .join('');
