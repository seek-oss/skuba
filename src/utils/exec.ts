import { cpus } from 'os';
import stream from 'stream';
import util from 'util';

import concurrently from 'concurrently';
import execa, { ExecaChildProcess } from 'execa';
import npmRunPath from 'npm-run-path';
import npmWhich from 'npm-which';

import { ConcurrentlyErrors, isErrorWithCode } from './error';
import { log } from './logging';

class YarnSpamFilter extends stream.Transform {
  silenced = false;

  _transform(
    chunk: any,
    _encoding: BufferEncoding,
    callback: stream.TransformCallback,
  ) {
    const str = Buffer.from(chunk).toString();

    // Yarn spews the entire installed dependency tree after this message
    if (str.startsWith('info Direct dependencies')) {
      this.silenced = true;
    }

    if (
      !this.silenced &&
      // This isn't very useful given the command generates a lockfile
      !str.startsWith('info No lockfile found')
    ) {
      this.push(chunk);
    }

    callback();
  }
}

class YarnWarningFilter extends stream.Transform {
  _transform(
    chunk: any,
    _encoding: BufferEncoding,
    callback: stream.TransformCallback,
  ) {
    const str = Buffer.from(chunk).toString();

    // Filter out annoying deprecation warnings that users can do little about
    if (!str.startsWith('warning skuba >')) {
      this.push(chunk);
    }

    callback();
  }
}

export type Exec = (
  command: string,
  ...args: string[]
) => ExecaChildProcess<string>;

interface ExecConcurrentlyCommand {
  command: string;
  name: string;
  prefixColor?: string;
}

interface ExecConcurrentlyOptions {
  /**
   * The maximum number of processes that can execute concurrently.
   *
   * Defaults to the CPU core count.
   */
  maxProcesses?: number;

  /**
   * A set length to pad names to.
   *
   * Defaults to the length of the longest command name.
   */
  nameLength?: number;

  /**
   * The stream that logging output will be written to.
   *
   * Defaults to `process.stdout`.
   */
  outputStream?: NodeJS.WritableStream;
}

type ExecOptions = execa.Options & { streamStdio?: true | 'yarn' };

const envWithPath = {
  PATH: npmRunPath({ cwd: __dirname }),
};

const runCommand = (command: string, args: string[], opts?: ExecOptions) => {
  const subprocess = execa(command, args, {
    localDir: __dirname,
    preferLocal: true,
    stdio: 'inherit',
    ...opts,
  });

  switch (opts?.streamStdio) {
    case 'yarn':
      const stderrFilter = new YarnWarningFilter();
      const stdoutFilter = new YarnSpamFilter();

      subprocess.stderr?.pipe(stderrFilter).pipe(process.stderr);
      subprocess.stdout?.pipe(stdoutFilter).pipe(process.stdout);

      break;

    case true:
      subprocess.stderr?.pipe(process.stderr);
      subprocess.stdout?.pipe(process.stdout);

      break;
  }

  return subprocess;
};

const whichCallback = npmWhich(__dirname);

const which = util.promisify<string, string>(whichCallback);

export const createExec =
  (opts: ExecOptions): Exec =>
  (command, ...args) =>
    runCommand(command, args, opts);

export const exec: Exec = (command, ...args) => runCommand(command, args);

export const execConcurrently = async (
  commands: ExecConcurrentlyCommand[],
  { maxProcesses, nameLength, outputStream }: ExecConcurrentlyOptions = {},
) => {
  const maxNameLength =
    nameLength ??
    commands.reduce(
      (length, command) => Math.max(length, command.name.length),
      0,
    );

  try {
    await concurrently(
      commands.map(({ command, name, prefixColor }) => ({
        command,
        env: envWithPath,
        name: name.padEnd(maxNameLength),
        prefixColor,
      })),
      {
        maxProcesses: maxProcesses ?? cpus().length,

        outputStream,

        // Use a minimalist logging prefix.
        prefix: '{name} |',
      },
    );
  } catch (err: unknown) {
    const result = ConcurrentlyErrors.validate(err);

    if (!result.success) {
      throw err;
    }

    const failed = result.value
      .filter(({ exitCode }) => exitCode !== 0)
      .sort(({ index: indexA }, { index: indexB }) => indexA - indexB)
      .map((subprocess) => subprocess.command.name);

    throw Error(
      `${failed.join(', ')} subprocess${
        failed.length === 1 ? '' : 'es'
      } failed.`,
    );
  }
};

export const ensureCommands = async (...names: string[]) => {
  let success = true;

  await Promise.all(
    names.map(async (name) => {
      const result = await hasCommand(name);

      if (!result) {
        success = false;

        log.err(log.bold(name), 'needs to be installed.');
      }
    }),
  );

  if (!success) {
    process.exit(1);
  }
};

export const hasCommand = async (name: string) => {
  try {
    await which(name);

    return true;
  } catch (err: unknown) {
    if (isErrorWithCode(err, 'ENOENT')) {
      return false;
    }

    throw err;
  }
};
