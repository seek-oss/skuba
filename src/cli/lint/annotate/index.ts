import type { ESLintOutput } from 'cli/adapter/eslint';
import type { PrettierOutput } from 'cli/adapter/prettier';

import type { StreamInterceptor } from '../external';

import { createBuildkiteAnnotations } from './buildkite';
import { createGitHubAnnotations } from './github';

export const createAnnotations = async (
  eslint: ESLintOutput,
  prettier: PrettierOutput,
  tscOk: boolean,
  tscOutputStream: StreamInterceptor,
): Promise<void> => {
  await Promise.all([
    createGitHubAnnotations(eslint, prettier, tscOk, tscOutputStream),
    createBuildkiteAnnotations(eslint, prettier, tscOk, tscOutputStream),
  ]);
};
