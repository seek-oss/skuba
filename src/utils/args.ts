import assert from 'assert';

import { COMMAND_ALIASES } from './command';

/**
 * Parse process arguments.
 *
 * This function mutates the input list by removing the command name element.
 * Downstream commands can access their args with an ordinary
 * `process.argv.slice(2)`.
 *
 * Example input:
 *
 * ```typescript
 * ['/bin/node', 'node_modules/.bin/skuba', 'test', '--foo', '--bar']
 * ```
 */
export const parseArgs = (args = process.argv) => {
  const skubaIdx = args.findIndex((chunk) => /skuba(\.[jt]s)?$/.test(chunk));

  assert(skubaIdx >= 0, 'Cannot parse args for `skuba`');

  const rawCommand = (args[skubaIdx + 1] ?? 'help').toLocaleLowerCase();

  const commandName = COMMAND_ALIASES[rawCommand] ?? rawCommand;

  const payload = {
    commandName,
    args: args.slice(skubaIdx + 2),
  };

  args.splice(skubaIdx + 1, 1);

  return payload;
};
