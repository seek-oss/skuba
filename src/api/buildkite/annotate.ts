import { exec, hasCommand } from '../../utils/exec';
import { log } from '../../utils/logging';

export type AnnotationStyle = 'success' | 'info' | 'warning' | 'error';

const isAnnotationEnabled = async () =>
  Boolean(
    process.env.BUILDKITE &&
      process.env.BUILDKITE_AGENT_ACCESS_TOKEN &&
      process.env.BUILDKITE_JOB_ID &&
      (await hasCommand('buildkite-agent')),
  );

interface AnnotationOptions {
  context?: string;

  /**
   * Scopes an annotation's context to the Buildkite step ID.
   *
   * This lets you emit distinct annotations per step, and only takes effect if
   * the `BUILDKITE_STEP_ID` environment variable is present.
   */
  scopeContextToStep?: boolean;

  style?: AnnotationStyle;
}

// Buildkite annotation currently only supports 1MiB of data
export const MAX_SIZE = 1024 * 1024; // 1MiB in bytes
export const TRUNCATION_WARNING = '... [Truncated due to size limit]';

/**
 * Asynchronously uploads a Buildkite annotation.
 *
 * If the following environment variables are not present,
 * the function will silently return without attempting to annotate:
 *
 * - `BUILDKITE`
 * - `BUILDKITE_AGENT_ACCESS_TOKEN`
 * - `BUILDKITE_JOB_ID`
 *
 * The `buildkite-agent` binary must also be on your `PATH`.
 */
export const annotate = async (
  markdown: string,
  opts: AnnotationOptions = {},
): Promise<void> => {
  if (!(await isAnnotationEnabled())) {
    return;
  }

  // Check if the annotation exceeds the maximum size
  let truncatedMarkdown = markdown;
  if (markdown.length > MAX_SIZE) {
    // Notify user of truncation, leave space for message
    const remainingSpace = MAX_SIZE - TRUNCATION_WARNING.length;
    truncatedMarkdown = markdown.slice(0, remainingSpace) + TRUNCATION_WARNING;
    // Log full message to the build log
    log.warn(`Annotation truncated, full message is: ${markdown}`);
  }

  // Always scope to the current Buildkite step.
  const context = [
    opts.scopeContextToStep && process.env.BUILDKITE_STEP_ID,
    opts.context,
  ]
    .filter(Boolean)
    .join('|');

  const { style } = opts;

  await exec(
    'buildkite-agent',
    'annotate',
    ...(context ? ['--context', context] : []),
    ...(style ? ['--style', style] : []),
    truncatedMarkdown,
  );
};
