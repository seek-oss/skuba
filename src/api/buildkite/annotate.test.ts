import * as execModule from '../../utils/exec';

import { annotate } from './annotate';

const exec = jest.spyOn(execModule, 'exec');
const hasCommand = jest.spyOn(execModule, 'hasCommand');

beforeEach(() => {
  jest.clearAllMocks();

  exec.mockResolvedValue(undefined as any);
  hasCommand.mockResolvedValue(true);
});

afterEach(() => {
  delete process.env.BUILDKITE;
  delete process.env.BUILDKITE_AGENT_ACCESS_TOKEN;
  delete process.env.BUILDKITE_JOB_ID;
  delete process.env.BUILDKITE_STEP_ID;
});

const setEnvironmentVariables = () => {
  process.env.BUILDKITE = 'true';
  process.env.BUILDKITE_AGENT_ACCESS_TOKEN = 'token';
  process.env.BUILDKITE_JOB_ID = 'job-id';
  process.env.BUILDKITE_STEP_ID = 'step-id';
};

describe('annotate', () => {
  const markdown = '**Message**';

  describe.each`
    description                    | opts
    ${'undefined options'}         | ${undefined}
    ${'empty options'}             | ${{}}
    ${'context option'}            | ${{ context: 'skuba-lint' }}
    ${'style option'}              | ${{ style: 'info' }}
    ${'context and style options'} | ${{ context: 'skuba-test', style: 'error' }}
    ${'all options'}               | ${{ context: 'skuba-build', scopeContextToStep: true, style: 'warning' }}
  `('with $description', ({ opts }) => {
    it('annotates when environment variables are set', async () => {
      setEnvironmentVariables();

      await expect(annotate(markdown, opts)).resolves.toBeUndefined();

      expect(
        [null, ...exec.mock.calls.flat(), null].join('\n'),
      ).toMatchSnapshot();
    });

    it('skips when environment variables are not set', async () => {
      await expect(annotate(markdown)).resolves.toBeUndefined();

      expect(exec).not.toHaveBeenCalled();
    });

    it('skips when `buildkite-agent` is not present', async () => {
      hasCommand.mockResolvedValue(false);

      await expect(annotate(markdown)).resolves.toBeUndefined();

      expect(exec).not.toHaveBeenCalled();
    });
  });
});
