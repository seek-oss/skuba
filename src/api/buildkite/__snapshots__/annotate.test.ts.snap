// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotate with all options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id|skuba-build
--style
warning
**Message**
"
`;

exports[`annotate with context and style options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
skuba-test
--style
error
**Message**
"
`;

exports[`annotate with context option annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
skuba-lint
**Message**
"
`;

exports[`annotate with empty options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
**Message**
"
`;

exports[`annotate with style option annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--style
info
**Message**
"
`;

exports[`annotate with undefined options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
**Message**
"
`;
